File: leetcode/Reverse Words in a String.cpp
class Solution {
public:
	void reverseWords(string &s) {
		size_t len = s.length();
		if (0 == len)
		{
			return;
		}
		for (size_t i = 0, j = 0; i <= len; ++i)
		{
			if (s.length() == i || ' ' == s[i] || '\t' == s[i] || '\n' == s[i] || '\r' == s[i])
			{
				if (i > j)
				{
					s += s.substr(j, i) + std::string(" ");
				}
				j = i;
			}
		}
		s = s.substr(len, s.length() - 1);
	}
};

File: answers/leetcode/Reverse Words in a String/Reverse Words in a String.java
public class Solution {
    public String reverseWords(String s) {
        if (null == s) {
            return null;
        }
        StringBuilder sb = new StringBuilder();
        for (int i = 0, j = 1; j <= s.length(); j++) {
            if (s.charAt(i) == ' ' && j < s.length() && s.charAt(j) != ' ') {
                if (i > 0) {
                    sb.insert(0, " ");
                }
                i = j;
            } else if (s.charAt(i) != ' ' && (j == s.length() || s.charAt(j) == ' ')) {
                sb.insert(0, s.substring(i, j));
                i = j;
            }
        }
        return sb.toString();
    }
}